stock_location_package_restriction/models/stock_location.py

```diff
@@ -16,21 +20,158 @@ class StockLocation(models.Model):
             Control if the location can contain products not in a package.
 
             Options:
-              * False (not set): Not mandatory, the location can contain products
-                not part of a package
+              * False (not set): No restriction, the location can contain products
+                with and without package
+              * Forbidden: The location cannot have products part of a package
               * Mandatory and unique: The location cannot have products not
                 part of a package and you cannot have more than 1 package on
                 the location
               * Mandatory and not unique:  The location cannot have products
-                not part of a package and you may have store multiple packages
+                not part of a package and you may store multiple packages
                 on the location
         """,
-        store=True,
     )
 
     @api.model
     def _selection_package_restriction(self):
         return [
+            (NOPACKAGE, "Forbidden"),
             (MULTIPACKAGE, "Mandatory"),
             (SINGLEPACKAGE, "Mandatory and unique"),
         ]
+
+    has_package_restriction_violation = fields.Boolean(
+        compute="_compute_has_package_restriction_violation",
+        search="_search_has_package_restriction_violation",
+        recursive=True,
+    )
+
+    def _has_package_restriction_violation_query(self):
+        self.flush()
+        query = """
+            SELECT stock_quant.location_id
+            FROM stock_quant
+            JOIN stock_location ON stock_location.id = stock_quant.location_id
+            WHERE
+                quantity != 0
+                AND stock_location.package_restriction IS NOT NULL
+            """
+        if self:
+            query += "AND stock_quant.location_id in %s"
+        query += f"""
+            GROUP BY
+                stock_quant.location_id, stock_location.package_restriction
+            HAVING
+                (
+                    stock_location.package_restriction = '{NOPACKAGE}'
+                    AND count(distinct(stock_quant.package_id)) > 0
+                ) OR (
+                    stock_location.package_restriction = '{SINGLEPACKAGE}'
+                    AND (
+                        count(distinct(stock_quant.package_id)) > 1
+                        OR count(*) FILTER (WHERE stock_quant.package_id IS NULL) > 0
+                    )
+                ) OR (
+                    stock_location.package_restriction = '{MULTIPACKAGE}'
+                    AND count(*) FILTER (WHERE stock_quant.package_id IS NULL) > 0
+                )
+        """
+        return query
```

Comment:
Would it not make sense to return the violation in sql?
Then we could display as well a violation string on the location and it could be reused by the `_check_package_restriction` method. 

```suggestion
        self.flush()
        nopackage_violation = f"stock_location.package_restriction = '{NOPACKAGE}' AND count(distinct(stock_quant.package_id)) > 0"
        singlepackage_violation = f"stock_location.package_restriction = '{SINGLEPACKAGE}' AND (count(distinct(stock_quant.package_id)) > 1 OR OR count(*) FILTER (WHERE stock_quant.package_id IS NULL) > 0)"
        multipackage_violation = f"stock_location.package_restriction = '{MULTIPACKAGE}' AND count(*) FILTER (WHERE stock_quant.package_id IS NULL) > 0"
        
        query = f"""
            SELECT
                stock_quant.location_id,
                CASE WHEN {nopackage_violation} THEN 1 ELSE 0 END AS nopackage_violation,
                CASE WHEN {singlepackage_violation} THEN 1 ELSE 0 END AS singlepackage_violation,
                CASE WHEN {multipackage_violation} THEN 1 ELSE 0 END AS multipackage_violation
            FROM stock_quant
            JOIN stock_location ON stock_location.id = stock_quant.location_id
            WHERE
                quantity != 0
                AND stock_location.package_restriction IS NOT NULL
            """
        if self:
            query += "AND stock_quant.location_id in %s"
        query += f"""
            GROUP BY
                stock_quant.location_id, stock_location.package_restriction
            HAVING
                ({nopackage_violation}) OR ({singlepackage_violation}) OR ({multipackage_violation})
        """
        return query
```